File: android/src/main/kotlin/com/example/pigeon_timer/PigeonTimerPlugin.kt
package com.example.pigeon_timer

import android.os.Build

import io.flutter.embedding.engine.plugins.FlutterPlugin


class PigeonTimerPlugin: FlutterPlugin, TimerHostApi {
  private var stopwatch : StopwatchTimer? = null;
  private var timerListenerApi: TimerListenerApi? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    TimerHostApi.setUp(binding.binaryMessenger, this)
    timerListenerApi = TimerListenerApi(binding.binaryMessenger)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    TimerHostApi.setUp(binding.binaryMessenger, null)
  }

  override fun getPlatformVersion(): String {
    return "Android " + Build.VERSION.RELEASE
  }

  override fun initTimer() {
    if (stopwatch != null) return;
    //Log.d("TAG","init native")
    stopwatch = StopwatchTimer()

    // Subscribe to the value change
    stopwatch?.onValueChanged = { newValue ->
      //return simple value
      timerListenerApi?.onTick(newValue.toLong()) {}
      // return as object class
      timerListenerApi?.onTickComplex(ComplexValue(
        newValue.toLong(),
        "Android"
      )) {}
    }
  }

  override fun start() {
    //Log.d("TAG","start native")
    stopwatch?.start();

  }

  override fun stop() {
   // Log.d("TAG","stop native")
    stopwatch?.stop();
  }

  override fun askImportantQuestion(question: Question, callback: (Result<Answer>) -> Unit) {
    val bigBrain = BigBrain()
    bigBrain.getAnswer = {
        answerText -> callback(Result.success(Answer(answerText)))
    }
    bigBrain.askQuestion(question.text)

  }
}

File: ios/Classes/PigeonTimerPlugin.swift
import Flutter
import UIKit


public class PigeonTimerPlugin: NSObject, FlutterPlugin, TimerHostApi {
    var stopwatch : StopwatchTimer? = nil
    static var timerListenerApi : TimerListenerApi? = nil
    
    public static func register(with registrar: any FlutterPluginRegistrar) {
        let messenger : FlutterBinaryMessenger = registrar.messenger()
        let api : TimerHostApi & NSObjectProtocol & FlutterPlugin = PigeonTimerPlugin.init()
        TimerHostApiSetup.setUp(binaryMessenger: messenger, api: api)
        PigeonTimerPlugin.timerListenerApi = TimerListenerApi(binaryMessenger: messenger)
        registrar.addApplicationDelegate(api)
    }
    
    
    func initTimer() throws {
        stopwatch = StopwatchTimer()
        
        stopwatch!.onValueChanged = { newValue in
            print("Counter updated: \(newValue)")
            PigeonTimerPlugin.timerListenerApi?.onTick(value: Int64(newValue), completion: { _ in })
            PigeonTimerPlugin.timerListenerApi?.onTickComplex(value: ComplexValue(value: Int64(newValue), platformName: "iOS"), completion: { _ in })
        }
    }
    
    func start() throws {
        if( stopwatch == nil){
            print("stopwatch is null")
            return;
        }
        stopwatch?.start()
    }
    
    func stop() throws {
        stopwatch?.stop()
    }
    
    func getPlatformVersion() throws -> String {
        return "iOS " + UIDevice.current.systemVersion;
    }
    
    func askImportantQuestion(question: Question, completion: @escaping (Result<Answer, any Error>) -> Void) {
        var bigBrain = BigBrain()
        bigBrain.getAnswer = { answerText in
            completion(Result.success(Answer(text: answerText)));
        }
        bigBrain.askQuestion(questionText: question.text)
    }
}
